File: multitogglebutton-java/src/main/java/com/dependencies/buria/multitogglebutton_java/MultiToggleButton.java
package com.dependencies.buria.multitogglebutton_java;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Typeface;
import android.util.AttributeSet;
import android.util.TypedValue;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.ImageButton;
import android.widget.LinearLayout;
import android.widget.TextView;


import java.util.ArrayList;
import java.util.Arrays;

public class MultiToggleButton extends LinearLayout {
    // typeface value
    private final int TEXT_STYLE_NORMAL = 0;
    private final int TEXT_STYLE_BOLD = 1;
    private final int TEXT_STYLE_ITALIC = 2;


    //attrs put user
    private ArrayList<Integer> toggleDrawables = null;
    private ArrayList<String> toggleStings = null;

    String text = "";
    Float buttonPadding = 0f;
    int toggleButtonTint = 0;
    Float textSize = 12f;
    int textColor =  0;
    private int textStyle = TEXT_STYLE_NORMAL;
    float toggleButtonSize = 24f;

    // variables counter
    private int currentItem = 0;
    private int itemCount = 0;
    private boolean isText = false;

    // listener
    private OnChangeListener onItemChangeListener = null;
    private OnTextChangeListener onTextChangeListener = null;

    //view
    TextView tvTgb;
    ImageButton tgb;
    private Context context;

    public MultiToggleButton(Context context) {
        super(context);
    }

    public MultiToggleButton(Context context, AttributeSet attrs) {
        super(context, attrs, 0);
        init(context, attrs, 0);
    }

    public MultiToggleButton(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        this.context = context;
        init(context, attrs, defStyleAttr);
    }

    public MultiToggleButton(Context context, AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
        init(context, attrs, defStyleAttr);
    }

    private void init(final Context context, AttributeSet attrs, int defStyleAttr){
        textSize = toDP(context, 12f);

        LayoutInflater inflater = (LayoutInflater) context.getSystemService(Context.LAYOUT_INFLATER_SERVICE);

        if (inflater != null)
            inflater.inflate(R.layout.layout_multi_toggle_button, this);

        tvTgb = findViewById(R.id.tvTgb);
        tgb = findViewById(R.id.tgb);
        TypedArray typeArray = context.obtainStyledAttributes(attrs, R.styleable.MultiToggleButton);


        try {
            text = typeArray.getString(R.styleable.MultiToggleButton_text);
            buttonPadding = typeArray.getDimension(R.styleable.MultiToggleButton_buttonPadding, toDP(context,8f));
            toggleButtonTint =
                    typeArray.getResourceId(R.styleable.MultiToggleButton_toggleButtonTint, 0);
            textSize = typeArray.getDimension(R.styleable.MultiToggleButton_textSize, toDP(context,14f));
            textColor = typeArray.getResourceId(R.styleable.MultiToggleButton_textColor, 0);
            textStyle = typeArray.getInt(R.styleable.MultiToggleButton_textStyle, TEXT_STYLE_NORMAL);
            toggleButtonSize = typeArray.getDimensionPixelSize(R.styleable.MultiToggleButton_toggleButtonSize,  Math.round(toDP(context,24f)));

            if (tvTgb != null) {
                tvTgb.setText(text);
                tvTgb.setPadding(buttonPadding.intValue(), 0, 0, 0);
                if (textColor != 0)
                    tvTgb.setTextColor(context.getResources().getColor(textColor));
                tvTgb.setTextSize(TypedValue.COMPLEX_UNIT_PX, textSize);

                if (textStyle != 0) {
                    switch (textStyle) {
                        case TEXT_STYLE_BOLD:
                            tvTgb.setTypeface(Typeface.DEFAULT_BOLD);
                            break;
                        case TEXT_STYLE_ITALIC:
                            tvTgb.setTypeface(tvTgb.getTypeface(), Typeface.ITALIC);
                            break;
                        default:
                            tvTgb.setTypeface(Typeface.DEFAULT);
                            break;
                    }
                }
            }

            if (tgb != null) {
                if (toggleButtonTint != 0)
                    tgb.setImageTintList(context.getResources().getColorStateList(toggleButtonTint));

                tgb.getLayoutParams().height = Math.round(toggleButtonSize);
                tgb.getLayoutParams().width = Math.round(toggleButtonSize);
            }
        }finally {
            typeArray.recycle();
        }

        if (tgb != null) {
            tgb.setOnClickListener(new OnClickListener() {
                @Override
                public void onClick(View v) {
                    setToggleImageResource();
                    setCurrentImageResource(context);

                    if (!isText) {
                        if (onItemChangeListener != null) {
                            onItemChangeListener.onItemChangeListener(getCurrentResource(),
                                    currentItem);
                        }
                    }else {
                        if (onTextChangeListener != null){
                            onTextChangeListener.onTextChangeListener(toggleStings.get(currentItem),
                                    currentItem);
                        }
                    }
                }
            });
        }

        if (tvTgb != null){
            tvTgb.setOnClickListener(new OnClickListener() {
                @Override
                public void onClick(View v) {
                    tgb.callOnClick();
                }
            });
        }
    }

    public void addToggleDrawables(int... resources) {
        isText = false;

        toggleDrawables = new ArrayList<>();
        if (resources != null) {
            for (int r : resources) {
                toggleDrawables.add(r);
            }
            itemCount = resources.length;
            setCurrentImageResource(context);
        }
    }

    public void addToggleTexts(String... resources) {
        isText = true;

        if (toggleStings == null)
            toggleStings = new ArrayList<>();

        if (resources != null) {
            toggleStings.addAll(Arrays.asList(resources));

            itemCount = resources.length;
            setCurrentImageResource(context);
        }
    }

    public void addToggleDrawables(ArrayList<Integer> resources) {
        isText = false;

        toggleDrawables = new ArrayList<>();
        if (resources != null) {
            toggleDrawables.addAll(resources);
            itemCount = resources.size();
            setCurrentImageResource(context);
        }
    }

    public void addToggleTexts(ArrayList<String> resources) {
        isText = true;

        if (toggleStings == null)
            toggleDrawables = new ArrayList<>();

        if (resources != null) {
            toggleStings.addAll(resources);
            itemCount = resources.size();
            setCurrentImageResource(context);
        }
    }

    private void setCurrentImageResource(Context context) {
        if (itemCount > 0){
            if (!isText) {
                if (toggleDrawables != null) {
                    if (currentItem < itemCount) {
                        if (toggleDrawables.get(currentItem) != null) {
                            tgb.setImageDrawable(context.getResources().getDrawable(toggleDrawables.get(currentItem)));
                        }
                    }
                }
            }else {
                if (toggleStings != null) {
                    if (currentItem < itemCount) {
                        if (toggleStings.get(currentItem) != null) {
                           tvTgb.setText(toggleStings.get(currentItem));
                        }
                    }
                }
            }
        }
    }

    private void setToggleImageResource() {
        currentItem = (currentItem == itemCount - 1) ? 0 : currentItem + 1;
    }

    private int getCurrentResource(){
        if (toggleDrawables != null){
            if (currentItem < itemCount){
                return toggleDrawables.get(currentItem);
            }
        }
        return 0;
    }


    public void setOnItemChangeListener(OnChangeListener onItemChangeListener) {
        this.onItemChangeListener = onItemChangeListener;
    }

    public void setOnItemChangeListener(OnTextChangeListener ontextChangeListener) {
        this.onTextChangeListener = ontextChangeListener;
    }

    public int getCurrentItem(){
        return getCurrentResource();
    }


    private Float toDP(Context context, Float value) {
        return TypedValue.applyDimension(
                TypedValue.COMPLEX_UNIT_DIP,
                value,
                context.getResources().getDisplayMetrics()
        );
    }

    private interface OnChangeListener{
        void onItemChangeListener(int resourceId, int position);
    }

    private interface OnTextChangeListener{
        void onTextChangeListener(String text, int position);
    }

    public void setDefaultPosition(int position){
        if (itemCount > 0)
            if (position >=0 && position < itemCount){
            currentItem = position;
            setCurrentImageResource(context);
        }
    }
}

File: multitogglebutton/src/main/java/com/dependencies/buria/multitogglebutton/MultiToggleButton.kt
package com.dependencies.buria.multitogglebutton

import android.content.Context
import android.content.res.TypedArray
import android.graphics.Typeface
import android.graphics.drawable.Drawable
import android.os.Build
import android.util.AttributeSet
import android.util.Log
import android.util.TypedValue
import android.view.LayoutInflater
import android.widget.LinearLayout
import android.widget.TextView
import androidx.core.view.marginLeft
import androidx.core.view.size
import kotlinx.android.synthetic.main.layout_multi_toggle_button.view.*
import kotlin.math.log

class MultiToggleButton @JvmOverloads constructor(
    context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0, defStyleRes: Int = 0
) : LinearLayout(context, attrs, defStyleAttr) {

    // typeface value
    private var TEXTSTYLE_NORMAL = 0
    private var TEXTSTYLE_BOLD = 1
    private var TEXTSTYLE_ITALIC = 2


    //attrs put user
    private var toggleDrawables: ArrayList<Int?>? = null
    private var toggleStrings: ArrayList<String>? = null
    var text: String? = ""
    var buttonPadding: Float = 0f
    var toggleButtonTint: Int = 0
    var textSize: Float = toDP(12f)
    var textColor: Int = 0
    private var textStyle: Int = TEXTSTYLE_NORMAL
    var toggleButtonSize: Float = 24f

    // variables counter
    private var currentItem: Int = 0
    private var itemCount: Int = 0
    private var isText: Boolean = false

    // listener
    private var onItemChangeListener: ((Int, Int) -> Unit)? = null
    private var onTextChangeListener: ((String, Int) -> Unit)? = null

    init {
        val inflater = context.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
        inflater.inflate(R.layout.layout_multi_toggle_button, this)
        val typeArray = context.obtainStyledAttributes(attrs, R.styleable.MultiToggleButton)

        try {
            text = typeArray.getString(R.styleable.MultiToggleButton_text)
            buttonPadding = typeArray.getDimension(R.styleable.MultiToggleButton_buttonPadding, toDP(8f))
            toggleButtonTint =
                typeArray.getResourceId(R.styleable.MultiToggleButton_toggleButtonTint, 0)
            textSize = typeArray.getDimension(R.styleable.MultiToggleButton_textSize, toDP(14f))
            textColor = typeArray.getResourceId(R.styleable.MultiToggleButton_textColor, 12)
            textStyle = typeArray.getInt(R.styleable.MultiToggleButton_textStyle,  TEXTSTYLE_NORMAL)
            toggleButtonSize = typeArray.getDimensionPixelSize(R.styleable.MultiToggleButton_toggleButtonSize,  toDP(24f).toInt()).toFloat()

            tvTgb.text = text
            tvTgb.setPadding(buttonPadding.toInt(), 0, 0, 0)
            tvTgb.setTextColor(context.resources.getColor(textColor))
            tvTgb.setTextSize(TypedValue.COMPLEX_UNIT_PX, textSize)

            when(textStyle){
                TEXTSTYLE_BOLD -> tvTgb.typeface = Typeface.DEFAULT_BOLD
                TEXTSTYLE_NORMAL -> tvTgb.typeface = Typeface.DEFAULT
                TEXTSTYLE_ITALIC -> tvTgb.setTypeface(tvTgb.typeface, Typeface.ITALIC)

                else -> tvTgb.typeface = Typeface.DEFAULT
            }

            if (toggleButtonTint != 0)
                tgb.imageTintList = context.resources.getColorStateList(toggleButtonTint)

            tgb.layoutParams.height = toggleButtonSize.toInt()
            tgb.layoutParams.width = toggleButtonSize.toInt()

        } finally {
            typeArray.recycle()
        }

        tgb.setOnClickListener {
            setToggleImageResource()
            setCurrentImageResource()
            if (!isText) {
                onItemChangeListener?.let { function ->
                    getCurrentResource()?.let { it1 -> function(it1, currentItem) }
                }
            } else{
                onTextChangeListener?.let { function ->
                    getCurrentText()?.let {
                        it -> function(it, currentItem)
                    }
                }
            }
        }

        tvTgb.setOnClickListener {
            tgb.callOnClick()
        }

    }



    fun addToggleDrawables(vararg resources: Int?) {
        isText = false
        toggleDrawables = arrayListOf()
        toggleDrawables?.addAll(resources)

        itemCount = resources.size
        setCurrentImageResource()
    }

    fun addToggleTexts(vararg resources: String) {
        isText = true
        toggleStrings = arrayListOf()
        toggleStrings?.addAll(resources)

        itemCount = resources.size
        setCurrentImageResource()
    }

    fun addToggleText(resources: ArrayList<String>?) {
        isText = true
        toggleStrings = arrayListOf()
        resources?.let { toggleStrings?.addAll(it) }

        itemCount = resources?.size ?: 0
        setCurrentImageResource()
    }

    fun addToggleTexts(resources: ArrayList<Int?>) {
        isText = false
        toggleDrawables = arrayListOf()
        toggleDrawables?.addAll(resources)

        itemCount = resources.size
        setCurrentImageResource()
    }

    private fun setCurrentImageResource() {
        if (itemCount > 0) {
            if (!isText)
                tgb.setImageDrawable(toggleDrawables?.get(currentItem)?.let { context.getDrawable(it) })
            else tvTgb.text = toggleStrings?.get(currentItem)
        }
    }

    private fun setToggleImageResource() {
        currentItem = if (currentItem == itemCount - 1) 0 else currentItem + 1
    }

    private fun getCurrentResource(): Int? {
        return toggleDrawables?.get(currentItem)
    }

    private fun getCurrentText(): String? {
        return toggleStrings?.get(currentItem)
    }

    fun setOnItemChangeListener(onItemChangeListener: (resourceId: Int, position: Int) -> Unit) {
        this.onItemChangeListener = onItemChangeListener
    }

    fun setOnTextChangeListener(onTextChangeListener: (text: String, position: Int) -> Unit){
        this.onTextChangeListener = onTextChangeListener
    }
    fun getCurrentItem(): Int? {
        return getCurrentResource()
    }

    private fun toDP(value: Float): Float {
        return TypedValue.applyDimension(
            TypedValue.COMPLEX_UNIT_DIP,
            value,
            context.resources.displayMetrics
        )
    }


    fun setDefaultPosition(position: Int){
        if (itemCount > 0)
            if (position in 0 until itemCount){
                currentItem = position
                setCurrentImageResource()
            }
    }


}